pkg/image/controller/trigger/image_trigger_controller.go

```diff
@@ -0,0 +1,437 @@
+	maxResourceInterval = 30 * time.Second
+)
+
+// ErrUnresolvedTag is used to indicate a resource is not ready to be triggered
+var ErrUnresolvedTag = fmt.Errorf("one or more triggers on this object cannot be resolved")
+
+// ImageStreamNamespaceLister helps get ImageStreams.
+// TODO: replace with generated informer interfaces
+type ImageStreamNamespaceLister interface {
+	// Get retrieves the Deployment from the indexer for a given namespace and name.
+	Get(name string, options metav1.GetOptions) (*imageapi.ImageStream, error)
+}
+
+// ImageStreamLister is the subset interface required off an ImageStream client to
+// implement this controller.
+// TODO: replace with generated informer interfaces
+type ImageStreamLister interface {
+	// ImageStreams returns an object that can get ImageStreams.
+	ImageStreams(namespace string) ImageStreamNamespaceLister
+}
+
+// TriggerSource defines the behavior for given resource type that can be triggered
+// by image stream tag changes.
+type TriggerSource struct {
+	// Resource is a value guaranteed to be unique across all TriggerSources passed
+	// to a given controller - used to separate keys processed by this controller.
+	Resource schema.GroupResource
+	// Informer is the source of changes for the resource type.
+	Informer cache.SharedInformer
+	// Store is the latest in memory cache for the resource type.
+	Store cache.Store
+	// TriggerFn must return a function that converts an object returned by the
+	// informer into a trigger.CacheEntry and add prefix to the trigger.CacheEntry key.
+	TriggerFn func(prefix string) trigger.Indexer
+	// Reactor is invoked when an image stream tag change is detected on one or
+	// more of the triggers defined on the object.
+	Reactor trigger.ImageReactor
+}
+
+// tagRetriever implements trigger.TagRetriever over an image stream lister.
+type tagRetriever struct {
+	lister ImageStreamLister
+}
+
+var _ trigger.TagRetriever = tagRetriever{}
+
+// NewTagRetriever will return a tag retriever that can look up image stream tag
+// references from an image stream.
+func NewTagRetriever(lister ImageStreamLister) trigger.TagRetriever {
+	return tagRetriever{lister}
+}
+
+// ImageStreamTag returns a valid image reference for the provided image stream tag name and namespace,
+// or returns false. rv is the resource version of the underlying image stream.
+func (r tagRetriever) ImageStreamTag(namespace, name string) (ref string, rv int64, ok bool) {
+	streamName, tag, ok := imageapi.SplitImageStreamTag(name)
+	if !ok {
+		return "", 0, false
+	}
+	is, err := r.lister.ImageStreams(namespace).Get(streamName, metav1.GetOptions{})
+	if err != nil {
+		return "", 0, false
+	}
+	rv, err = strconv.ParseInt(is.ResourceVersion, 10, 64)
+	if err != nil {
+		return "", 0, false
+	}
+	ref, ok = imageapi.ResolveLatestTaggedImage(is, tag)
+	return ref, rv, ok
+}
+
+// StoreToImageStreamLister adapts a cache.Store to ImageStreamLister
+// TODO: replace with generated informer interfaces
```

Comment:
It's a bigger deal, we need to switch to generated informers in the first place to be able to use those. I've opened https://github.com/openshift/origin/issues/14108 to address that.